### ISSUE TITLE
CMS 4.12 regression: Incorrect template shown when user doesn’t have CMS access

### ISSUE THREAD

**Issue by lozcalver** [silverstripe/silverstripe-admin#1443]
Likely a regression from https://github.com/silverstripe/silverstripe-admin/pull/1317

If a member without any CMS access permissions is logged in and visits /admin, they will see a blank admin area and a generic error message:

![Screenshot 2023-01-17 at 10 40 04](https://user-images.githubusercontent.com/1655548/213130235-b1d5f764-46c5-420f-9324-8a2da173455f.png)

Prior to admin 1.12, the user would be shown a frontend template with a more descriptive error message:

![Screenshot 2023-01-17 at 10 39 48](https://user-images.githubusercontent.com/1655548/213130648-e799fecb-d6de-4136-8c85-7228c419d5e2.png)

## Acceptance criteria
- Users with no CMS access should not see errors in CMS context.
- ~If there's an 403 error page define, users without CMS access who try to access the CMS should see that error page.~ - this was not the old behaviour, and is actually _less_ useful than the old behaviour which gives you a "log in as someone else" button in case you need to log in as a different user.
- If there's no error page, the previous behaviour is restored.

## Note
- `AdminRootController` loops through all the LeftAndMain. At first glance it would make sense to double check if the Member as access to the CMS before looping through any of those section.
- Traditionally, we consider you have CMS Access if you have any `CMS_ADMIN_*` permission.

## PRs
- https://github.com/silverstripe/silverstripe-admin/pull/1466

**Comment by lekoala** [comment]
i have the same issue, very annoying

I've fixed it so far by added a statuscode check in afterHandleRequest

```diff
 public function afterHandleRequest()
 {
     if (
         $this->response->isError()
+        &&  $this->response->getStatusCode() !== 403
         && !$this->request->isAjax()
         && $this->response->getHeader('Content-Type') !== 'application/json'
     ) {
```

**Comment by elliot-sawyer** [comment]
I tried @lekoala 's fix by subclassing LeftAndMain and replacing with Injector, but it doesn't seem to call the `afterHandleRequest()`.  LeftAndMain subclasses themselves require a $url_segment and appear in the admin menu, so the approach isn't viable. 

It seems a direct edit to LeftAndMain.php is required to fix, so devs would need to fork `framework` to fix this.

My attempted fix:

```php
class MyLeftAndMain extends LeftAndMain
{
    /**
     * Identical to the parent, except for `$this->response->getStatusCode() !== 403`
     *
     * @return void
     */
    public function afterHandleRequest()
    {
        if ($this->response->isError()
            && $this->response->getStatusCode() !== 403
            && !$this->request->isAjax()
            && $this->response->getHeader('Content-Type') !== 'application/json'
        ) {
            $this->init();
            $errorCode = $this->response->getStatusCode();
            $errorType = $this->response->getStatusDescription();
            $defaultMessage = _t(
                self::class . '.ErrorMessage',
                'Sorry, it seems there was a {errorcode} error.',
                ['errorcode' => $errorCode]
            );
            $this->response = HTTPResponse::create($this->render([
                'Title' => $this->getApplicationName() . ' - ' . $errorType,
                'Content' => $this->renderWith($this->getTemplatesWithSuffix('_Error'), [
                    'ErrorCode' => $errorCode,
                    'ErrorType' => $errorType,
                    'Message' => DBField::create_field(
                        'HTMLFragment',
                        _t(self::class . '.ErrorMessage' . $errorCode, $defaultMessage)
                    ),
                ]),
            ]), $errorCode, $errorType);
        }
        parent::afterHandleRequest();
    }
}
```

```yml
SilverStripe\Core\Injector\Injector:
  SilverStripe\Admin\LeftAndMain:
    class: Namespace\To\MyLeftAndMain
```

**Comment by GuySartorelli** [comment]
Sounds like in addition to fixing the regression, this new error handling behaviour should maybe be made customisable (probably with extension hooks, and potentially also by adding a key to the extension yml, which arguably should be a requirement for all extensions added via yaml in core modules)

**Comment by GuySartorelli** [comment]
@elliot-sawyer Part of why the solution you attempted won't work is because subclasses of `LeftAndMain` aren't subclasses of your new injected class - so they're all still using their original implementations of the method.

**Comment by lekoala** [comment]
@elliot-sawyer yes my fix needs to be applied on core code
@GuySartorelli do you think my approach would fix the issue? i think it works decently enough, better than what's currently in place anyway.

**Comment by GuySartorelli** [comment]
I don't think just blanket ignoring 403 is correct without checking further - I think things like "Action 'SomeAction' isn't available on class Some\Class" are 403, and we don't want those kicking to the frontend. That's actually one of the main errors that made me jump on making the change in the first place.

**Comment by michalkleiner** [comment]
> I don't think just blanket ignoring 403 is correct without checking further - I think things like "Action 'SomeAction' isn't available on class Some\Class" are 403, and we don't want those kicking to the frontend. That's actually one of the main errors that made me jump on making the change in the first place.

At that point we probably already have an existing user/session, right? You won't get that from the CMS unless you were in the CMS before.. and if you weren't logged into the CMS then ending on the front-end is fine.

**Comment by GuySartorelli** [comment]
This issue is reported as happening when you _do_ have an authenticated user and session, but they just don't have any permissions for `LeftAndMain` sections.

**Comment by michalkleiner** [comment]
Ah, true, mixed it up with a member not existing at all situation.

**Comment by lozcalver** [comment]
(Thinking aloud here, I haven’t dug into this at all)

So it sounds to me that we now need a way to opt-in to theme-based errors on a case-by-case basis. I.e. in core code we can indicate, for specific calls to `Security::permissionFailure()`, that they should be rendered with a front-end template and not use the admin theme at all.

Could it be as simple as calling `SSViewer::set_themes(SSViewer::config()->get('themes'))` before we call `Security::permissionFailure()` for those cases?

Edit: I’ve just noticed `AdminErrorExtension`, so perhaps an API to opt out of that extension’s error handling?

```php
AdminErrorExtension::withFrontendTemplates(function() {
    Security::permissionFailure();
});
```

**Comment by lekoala** [comment]
Basically, the logic should be: for unauthorized users, you should not see the admin layout.
Currently, i'm checking this using 403 responses in my fix. Other approaches may work as well. I don't think there is a need to change how permissionFailure works.

**Comment by michalkleiner** [comment]
@lekoala I think the problem here is that we need to correctly cater for two separate scenarios, where one is a user without a member record at all, and a user with a member record but with no permissions to view anything. Should those be treated the same or differently?

**Comment by elliot-sawyer** [comment]
@michalkleiner  I arrive at this scenario for authenticated users who have no CMS permissions.  Someone without a Member record would never be able to log in

**Comment by lekoala** [comment]
@michalkleiner in my view, if you have CMS_ACCESS permission, you can view the admin layout, otherwise, you should view the regular layout. So maybe, on top of my current fix, we can check also for the current user and their permission.

**Comment by michalkleiner** [comment]
Yep. Also consider the situation the framework with its modeladmins is used separately without the CMS, it still provides the admin interface where this can happen as well.

**Comment by GuySartorelli** [comment]
This has been fixed and released as [1.12.3](https://github.com/silverstripe/silverstripe-admin/releases/tag/1.12.3).

**Comment by lekoala** [comment]
I'm sorry to revive this, but with subsites, this doesn't work as expected

due to this 

https://github.com/silverstripe/silverstripe-subsites/blob/6bcc688c78680098c69afcbe82e1a45ae3081374/src/Extensions/LeftAndMainSubsites.php#L371

and this

        // Allow customisation of the access check by a extension
        // Also all the canView() check to execute Controller::redirect()
        if (!$this->canView() && !$this->getResponse()->isFinished()) {

the subsite extension runs before, and throw a permissionFailure

not sure what needs to be fixed. i wouldn't allow extension to customize anything, and simply do

        if (!$this->canView())) {

which seems safer anyway
